include/dsn/internal/network.h

```diff
@@ -286,16 +287,13 @@ namespace dsn {
         uint64_t                           _message_sent;
         // ]
 
-        ::dsn::utils::ex_lock_nr           _delay_lock; // [
-        int                                _delay_server_receive_ms;
-        // ]
+        std::atomic_int                    _delay_server_receive_ms;
     };
 
     // --------- inline implementation --------------
     inline void rpc_session::delay_recv(int delay_ms)
     {
-        utils::auto_lock<utils::ex_lock_nr> l(_delay_lock);
-        if (delay_ms > _delay_server_receive_ms)
-            _delay_server_receive_ms = delay_ms;
+        int old_delay_ms = _delay_server_receive_ms.load();
+        while (delay_ms > old_delay_ms && !_delay_server_receive_ms.compare_exchange_weak(old_delay_ms, delay_ms)) {}
```

Comment:
Refer to http://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange:
```
bool compare_exchange_weak( T& expected, T desired,
                            std::memory_order success,
                            std::memory_order failure );

Atomically compares the object representation of *this with the object representation of expected, as if by std::memcmp, and if those are bitwise-equal, replaces the former with desired (performs read-modify-write operation). Otherwise, loads the actual value stored in *this into expected (performs load operation). Copying is performed as if by std::memcpy. 
```